File: benchmarks/multilang-wasm/text-gc-document-edit/gc_document_kernel.c
// gc_document_kernel.c — multilang compute core for text.gc-document-edit.v1.
//
// ABI (mirrors the multilang-wasm kernels): the adapter writes the frozen
// fixture bytes into linear memory at offset 0 (byte length passed in), and
// the kernel parses parseFixture + executeFixture (256 initial labelled nodes
// on a 3-ary DAG plus 10,000 edits — inserts/deletes/reparents mirroring
// benchmarks/v1/text-gc-document-edit/workload.js), writes counters + a
// deterministic FNV-1a digest of the DFS canonical traversal (u32 id +
// label bytes + u32 child count per node) to a fixed memory offset, and
// returns the final node count.
//
// Results (fixed offset 327680):
//   [0] inserts             (=3334)   [1] deletes         (=3333)
//   [2] reparents           (=3333)   [3] final-nodes     (=257)
//   [4] child-insertions    (=6922)   [5] child-removals  (=6666)
//   [6] parent-writes       (=10255)  [7] canonical FNV-1a (u32)
// Exports: i32 gc_document_edit_trace(u32 fixture_len) -> final_nodes

#define MAX_SLOTS 4096                 // ids in the frozen fixture max at 3589
// clang/wasm-ld defaults place the stack at 0..64KiB and the .bss/data
// section immediately after (data_end ≈ 172 KiB for our slot arrays). Put
// the fixture bytes and results well above data_end so the adapter's write
// cannot clobber our slot storage. Compile with --initial-memory=655360.
#define FIXTURE_OFFSET 196608          // 192 KiB, past LLD's default data_end
#define RES_OFFSET 524288              // 512 KiB — past 256 KiB max fixture

typedef unsigned int u32;
typedef int i32;
typedef unsigned short u16;
typedef unsigned char u8;

// Slot storage (indexed by fixture id, which is <= 3589)
static i32 parent_of[MAX_SLOTS];       // -2 = absent, -1 = root, else parent id
static i32 first_child_of[MAX_SLOTS];  // -1 = none
static i32 prev_sibling_of[MAX_SLOTS]; // -1 = first child
static i32 next_sibling_of[MAX_SLOTS]; // -1 = last child
static u32 child_count_of[MAX_SLOTS];
static u32 label_off_of[MAX_SLOTS];    // hex label byte offset in fixture
static u16 label_hex_len_of[MAX_SLOTS];

// FNV-1a 32-bit
static u32 fnv;
static void fnv_reset(void) { fnv = 0x811c9dc5U; }
static void fnv_mix_byte(u8 b) { fnv ^= (u32)b; fnv *= 0x01000193U; }
static void fnv_mix_u32(u32 v) {
  fnv_mix_byte((u8)(v & 0xff));
  fnv_mix_byte((u8)((v >> 8) & 0xff));
  fnv_mix_byte((u8)((v >> 16) & 0xff));
  fnv_mix_byte((u8)((v >> 24) & 0xff));
}

static u8 fixture_at(u32 off) { return *((u8 *)(FIXTURE_OFFSET) + off); }
static int is_digit(u8 c) { return c >= '0' && c <= '9'; }
static int is_hex(u8 c) {
  return (c >= '0' && c <= '9') || (c >= 'a' && c <= 'f') || (c >= 'A' && c <= 'F');
}
static u8 hex_val(u8 c) {
  if (c >= '0' && c <= '9') return c - '0';
  if (c >= 'a' && c <= 'f') return c - 'a' + 10;
  return c - 'A' + 10;
}

// Read decimal integer at *off; may be negative (with leading '-').
// Advances *off past the number. Returns the value.
static i32 read_int(u32 *off, u32 end) {
  int neg = 0;
  if (*off < end && fixture_at(*off) == '-') { neg = 1; (*off)++; }
  i32 v = 0;
  while (*off < end && is_digit(fixture_at(*off))) {
    v = v * 10 + (i32)(fixture_at(*off) - '0');
    (*off)++;
  }
  return neg ? -v : v;
}
// Advance *off past a single delimiter (\t or \n).
static void skip_delim(u32 *off) { (*off)++; }
// Read the hex-label span at *off up to \t or \n; return start_off & hex_len.
static void read_hex_span(u32 *off, u32 end, u32 *start_off, u16 *hex_len) {
  *start_off = *off;
  u32 start = *off;
  while (*off < end && fixture_at(*off) != '\t' && fixture_at(*off) != '\n') {
    (*off)++;
  }
  *hex_len = (u16)(*off - start);
}
// Skip to the next newline (or end).
static void skip_line(u32 *off, u32 end) {
  while (*off < end && fixture_at(*off) != '\n') (*off)++;
  if (*off < end) (*off)++;
}
// Skip past a header line by finding a \t, reading the count, then skipping to newline.
static i32 read_header_count(u32 *off, u32 end) {
  while (*off < end && fixture_at(*off) != '\t') (*off)++;
  if (*off < end) (*off)++;
  i32 count = read_int(off, end);
  skip_line(off, end);
  return count;
}

// Doubly linked list helpers.
static void link_after(i32 parent, i32 anchor, i32 node) {
  // insert `node` immediately after `anchor` inside parent's children.
  if (anchor == -1) {
    // anchor == -1 means insert at head
    const i32 old_head = first_child_of[parent];
    next_sibling_of[node] = old_head;
    prev_sibling_of[node] = -1;
    if (old_head != -1) prev_sibling_of[old_head] = node;
    first_child_of[parent] = node;
  } else {
    const i32 old_next = next_sibling_of[anchor];
    next_sibling_of[node] = old_next;
    prev_sibling_of[node] = anchor;
    if (old_next != -1) prev_sibling_of[old_next] = node;
    next_sibling_of[anchor] = node;
  }
  child_count_of[parent]++;
}

static void insert_at_position(i32 parent, i32 position, i32 node) {
  // walk to find the (position-1)th child. Anchor = -1 when position == 0.
  if (position == 0) {
    link_after(parent, -1, node);
    return;
  }
  i32 cur = first_child_of[parent];
  for (i32 k = 0; k < position - 1 && cur != -1; k++) cur = next_sibling_of[cur];
  link_after(parent, cur, node);
}

static void splice_out(i32 node) {
  const i32 par = parent_of[node];
  const i32 p = prev_sibling_of[node];
  const i32 n = next_sibling_of[node];
  if (p == -1) first_child_of[par] = n;
  else next_sibling_of[p] = n;
  if (n != -1) prev_sibling_of[n] = p;
  prev_sibling_of[node] = -1;
  next_sibling_of[node] = -1;
  child_count_of[par]--;
}

// Decode a hex-encoded label and feed each decoded byte to FNV.
static u32 mix_label(u32 off, u16 hex_len) {
  const u32 byte_len = (u32)(hex_len / 2);
  fnv_mix_u32(byte_len);
  for (u32 i = 0; i < byte_len; i++) {
    const u8 hi = fixture_at(off + i * 2);
    const u8 lo = fixture_at(off + i * 2 + 1);
    fnv_mix_byte((u8)((hex_val(hi) << 4) | hex_val(lo)));
  }
  return byte_len;
}

static void dfs_mix(i32 slot) {
  fnv_mix_u32((u32)slot);
  mix_label(label_off_of[slot], label_hex_len_of[slot]);
  fnv_mix_u32(child_count_of[slot]);
  for (i32 c = first_child_of[slot]; c != -1; c = next_sibling_of[c]) {
    dfs_mix(c);
  }
}

__attribute__((export_name("gc_document_edit_trace")))
int gc_document_edit_trace(u32 fixture_len) {
  u32 *results = (u32 *)RES_OFFSET;
  for (i32 i = 0; i < MAX_SLOTS; i++) {
    parent_of[i] = -2;
    first_child_of[i] = -1;
    prev_sibling_of[i] = -1;
    next_sibling_of[i] = -1;
    child_count_of[i] = 0;
    label_off_of[i] = 0;
    label_hex_len_of[i] = 0;
  }

  u32 off = 0;
  const u32 end = fixture_len;
  // Skip format-header line.
  skip_line(&off, end);
  // "initial\t<N>\n"
  const i32 initial_count = read_header_count(&off, end);
  // "operations\t<M>\n"
  (void)read_header_count(&off, end);

  u32 child_insertions = 0;
  u32 child_removals = 0;
  u32 parent_writes = 0;
  u32 node_count = 0;

  // Read initial rows: "N\t<id>\t<parentId>\t<position>\t<hexLabel>\n"
  for (i32 i = 0; i < initial_count; i++) {
    // 'N' then '\t'
    off += 2;
    const i32 id = read_int(&off, end);
    skip_delim(&off);
    const i32 parentId = read_int(&off, end);
    skip_delim(&off);
    const i32 position = read_int(&off, end);
    skip_delim(&off);
    u32 label_off = 0;
    u16 label_hex_len = 0;
    read_hex_span(&off, end, &label_off, &label_hex_len);
    skip_line(&off, end);

    label_off_of[id] = label_off;
    label_hex_len_of[id] = label_hex_len;
    if (parentId == -1) {
      // root
      parent_of[id] = -1;
      first_child_of[id] = -1;
      child_count_of[id] = 0;
    } else {
      parent_of[id] = parentId;
      first_child_of[id] = -1;
      child_count_of[id] = 0;
      insert_at_position(parentId, position, id);
      child_insertions++;
      parent_writes++;
    }
    node_count++;
  }

  u32 inserts = 0, deletes = 0, reparents = 0;

  // Ops loop
  while (off < end) {
    const u8 tag = fixture_at(off);
    if (tag == '\n') { off++; continue; }
    off++;               // skip tag
    skip_delim(&off);    // skip \t
    if (tag == 'I') {
      const i32 id = read_int(&off, end); skip_delim(&off);
      const i32 parentId = read_int(&off, end); skip_delim(&off);
      const i32 position = read_int(&off, end); skip_delim(&off);
      u32 label_off = 0;
      u16 label_hex_len = 0;
      read_hex_span(&off, end, &label_off, &label_hex_len);
      skip_line(&off, end);

      label_off_of[id] = label_off;
      label_hex_len_of[id] = label_hex_len;
      parent_of[id] = parentId;
      first_child_of[id] = -1;
      child_count_of[id] = 0;
      insert_at_position(parentId, position, id);
      inserts++;
      child_insertions++;
      parent_writes++;
      node_count++;
    } else if (tag == 'D') {
      const i32 id = read_int(&off, end);
      skip_line(&off, end);
      splice_out(id);
      parent_of[id] = -2;
      deletes++;
      child_removals++;
      parent_writes++;
      node_count--;
    } else if (tag == 'R') {
      const i32 id = read_int(&off, end); skip_delim(&off);
      const i32 parentId = read_int(&off, end); skip_delim(&off);
      const i32 position = read_int(&off, end);
      skip_line(&off, end);
      splice_out(id);
      parent_of[id] = parentId;
      insert_at_position(parentId, position, id);
      reparents++;
      child_insertions++;
      child_removals++;
      parent_writes++;
    } else {
      // unknown tag — skip line defensively
      skip_line(&off, end);
    }
  }

  // Traversal-based canonical FNV-1a digest.
  fnv_reset();
  dfs_mix(0);
  const u32 canonical_fnv = fnv;

  results[0] = inserts;
  results[1] = deletes;
  results[2] = reparents;
  results[3] = node_count;
  results[4] = child_insertions;
  results[5] = child_removals;
  results[6] = parent_writes;
  results[7] = canonical_fnv;
  return (int)node_count;
}

File: benchmarks/multilang-wasm/text-gc-document-edit/gc_document_kernel.ts
// gc_document_kernel.ts — AssemblyScript multilang compute core for
// text.gc-document-edit.v1. Same ABI as gc_document_kernel.c: the adapter
// writes the frozen fixture bytes into linear memory at FIXTURE_OFFSET
// (byte length passed in), and the kernel parses parseFixture + executeFixture
// (256 initial labelled nodes on a 3-ary DAG plus 10,000 edits — inserts,
// leaf-deletes, and reparents mirroring benchmarks/v1/text-gc-document-edit/
// workload.js), writes counters + a deterministic FNV-1a digest of the DFS
// canonical traversal (u32 id + label bytes + u32 child count per node) to
// RES_OFFSET, and returns the final node count. Raw linear-memory access
// only (no heap allocation, no runtime imports) — mirrors grid_kernel.ts.

const MAX_SLOTS: i32 = 4096;
// Slot-storage layout (all before the fixture window at 196608):
//   PARENT_OF:        i32[4096] at 0       (bytes 0..16384)
//   FIRST_CHILD_OF:   i32[4096] at 16384   (bytes 16384..32768)
//   PREV_SIBLING_OF:  i32[4096] at 32768   (bytes 32768..49152)
//   NEXT_SIBLING_OF:  i32[4096] at 49152   (bytes 49152..65536)
//   CHILD_COUNT_OF:   u32[4096] at 65536   (bytes 65536..81920)
//   LABEL_OFF_OF:     u32[4096] at 81920   (bytes 81920..98304)
//   LABEL_HEX_LEN_OF: u16[4096] at 98304   (bytes 98304..106496)
const PARENT_OF: usize = 0;
const FIRST_CHILD_OF: usize = 16384;
const PREV_SIBLING_OF: usize = 32768;
const NEXT_SIBLING_OF: usize = 49152;
const CHILD_COUNT_OF: usize = 65536;
const LABEL_OFF_OF: usize = 81920;
const LABEL_HEX_LEN_OF: usize = 98304;
const FIXTURE_OFFSET: usize = 196608; // 192 KiB
const RES_OFFSET: usize = 524288; // 512 KiB

let fnv: u32 = 0;

// --- FNV-1a 32-bit ---------------------------------------------------------
function fnvReset(): void {
  fnv = 0x811c9dc5;
}
function fnvMixByte(b: u8): void {
  fnv = ((fnv ^ (<u32> b)) * 0x01000193) >>> 0;
}
function fnvMixU32(v: u32): void {
  fnvMixByte(<u8> (v & 0xff));
  fnvMixByte(<u8> ((v >>> 8) & 0xff));
  fnvMixByte(<u8> ((v >>> 16) & 0xff));
  fnvMixByte(<u8> ((v >>> 24) & 0xff));
}

// --- Slot accessors --------------------------------------------------------
function getParent(id: i32): i32 {
  return load<i32>(PARENT_OF + (<usize> id) * 4);
}
function setParent(id: i32, v: i32): void {
  store<i32>(PARENT_OF + (<usize> id) * 4, v);
}
function getFirstChild(id: i32): i32 {
  return load<i32>(FIRST_CHILD_OF + (<usize> id) * 4);
}
function setFirstChild(id: i32, v: i32): void {
  store<i32>(FIRST_CHILD_OF + (<usize> id) * 4, v);
}
function getPrevSibling(id: i32): i32 {
  return load<i32>(PREV_SIBLING_OF + (<usize> id) * 4);
}
function setPrevSibling(id: i32, v: i32): void {
  store<i32>(PREV_SIBLING_OF + (<usize> id) * 4, v);
}
function getNextSibling(id: i32): i32 {
  return load<i32>(NEXT_SIBLING_OF + (<usize> id) * 4);
}
function setNextSibling(id: i32, v: i32): void {
  store<i32>(NEXT_SIBLING_OF + (<usize> id) * 4, v);
}
function getChildCount(id: i32): u32 {
  return load<u32>(CHILD_COUNT_OF + (<usize> id) * 4);
}
function setChildCount(id: i32, v: u32): void {
  store<u32>(CHILD_COUNT_OF + (<usize> id) * 4, v);
}
function getLabelOff(id: i32): u32 {
  return load<u32>(LABEL_OFF_OF + (<usize> id) * 4);
}
function setLabelOff(id: i32, v: u32): void {
  store<u32>(LABEL_OFF_OF + (<usize> id) * 4, v);
}
function getLabelHexLen(id: i32): u16 {
  return load<u16>(LABEL_HEX_LEN_OF + (<usize> id) * 2);
}
function setLabelHexLen(id: i32, v: u16): void {
  store<u16>(LABEL_HEX_LEN_OF + (<usize> id) * 2, v);
}

// --- Fixture byte accessors ------------------------------------------------
function fixtureAt(off: u32): u8 {
  return load<u8>(FIXTURE_OFFSET + (<usize> off));
}
function isDigit(c: u8): bool {
  return c >= 0x30 && c <= 0x39;
}
function hexVal(c: u8): u8 {
  if (c >= 0x30 && c <= 0x39) return c - 0x30;
  if (c >= 0x61 && c <= 0x66) return c - 0x61 + 10;
  return c - 0x41 + 10;
}

// --- Parser primitives -----------------------------------------------------
function readInt(offBase: u32, end: u32): u64 {
  // Encodes (value:i32 lo32, newOffset:u32 hi32) into a u64 for a single return.
  let off = offBase;
  let neg = false;
  if (off < end && fixtureAt(off) == 0x2d /* '-' */) {
    neg = true;
    off++;
  }
  let v: i32 = 0;
  while (off < end && isDigit(fixtureAt(off))) {
    v = v * 10 + (<i32> (fixtureAt(off) - 0x30));
    off++;
  }
  if (neg) v = -v;
  return (<u64> off << 32) | (<u64> (<u32> v));
}
function skipLine(offBase: u32, end: u32): u32 {
  let off = offBase;
  while (off < end && fixtureAt(off) != 0x0a /* '\n' */) off++;
  if (off < end) off++;
  return off;
}
function readHeaderCount(offBase: u32, end: u32): u64 {
  let off = offBase;
  while (off < end && fixtureAt(off) != 0x09 /* '\t' */) off++;
  if (off < end) off++;
  const packed = readInt(off, end);
  off = <u32> (packed >>> 32);
  const c = <i32> (packed & 0xffffffff);
  off = skipLine(off, end);
  return (<u64> off << 32) | (<u64> (<u32> c));
}
// readHexSpan returns (start_off:u32 lo, hex_len:u16, newOff:u32 in hi32)
// We split into two calls to keep u64 packing simple: readHexSpanStart / End.
function readHexSpanEnd(offBase: u32, end: u32): u32 {
  let off = offBase;
  while (
    off < end && fixtureAt(off) != 0x09 /* '\t' */ && fixtureAt(off) != 0x0a /* '\n' */
  ) {
    off++;
  }
  return off;
}

// --- List/tree helpers -----------------------------------------------------
function linkAfter(parent: i32, anchor: i32, node: i32): void {
  if (anchor == -1) {
    const oldHead = getFirstChild(parent);
    setNextSibling(node, oldHead);
    setPrevSibling(node, -1);
    if (oldHead != -1) setPrevSibling(oldHead, node);
    setFirstChild(parent, node);
  } else {
    const oldNext = getNextSibling(anchor);
    setNextSibling(node, oldNext);
    setPrevSibling(node, anchor);
    if (oldNext != -1) setPrevSibling(oldNext, node);
    setNextSibling(anchor, node);
  }
  setChildCount(parent, getChildCount(parent) + 1);
}
function insertAtPosition(parent: i32, position: i32, node: i32): void {
  if (position == 0) {
    linkAfter(parent, -1, node);
    return;
  }
  let cur = getFirstChild(parent);
  let k: i32 = 0;
  while (k < position - 1 && cur != -1) {
    cur = getNextSibling(cur);
    k++;
  }
  linkAfter(parent, cur, node);
}
function spliceOut(node: i32): void {
  const par = getParent(node);
  const p = getPrevSibling(node);
  const n = getNextSibling(node);
  if (p == -1) setFirstChild(par, n);
  else setNextSibling(p, n);
  if (n != -1) setPrevSibling(n, p);
  setPrevSibling(node, -1);
  setNextSibling(node, -1);
  setChildCount(par, getChildCount(par) - 1);
}

function dfsMix(slot: i32): void {
  fnvMixU32(<u32> slot);
  const hexLen = <u32> getLabelHexLen(slot);
  const byteLen = hexLen / 2;
  fnvMixU32(byteLen);
  const off = getLabelOff(slot);
  for (let i: u32 = 0; i < byteLen; i++) {
    const hi = fixtureAt(off + i * 2);
    const lo = fixtureAt(off + i * 2 + 1);
    fnvMixByte(<u8> ((hexVal(hi) << 4) | hexVal(lo)));
  }
  fnvMixU32(getChildCount(slot));
  let c = getFirstChild(slot);
  while (c != -1) {
    dfsMix(c);
    c = getNextSibling(c);
  }
}

export function gc_document_edit_trace(fixture_len: u32): i32 {
  // Reset slot storage.
  for (let i: i32 = 0; i < MAX_SLOTS; i++) {
    setParent(i, -2);
    setFirstChild(i, -1);
    setPrevSibling(i, -1);
    setNextSibling(i, -1);
    setChildCount(i, 0);
    setLabelOff(i, 0);
    setLabelHexLen(i, 0);
  }

  let off: u32 = 0;
  const end = fixture_len;
  off = skipLine(off, end);
  let hc = readHeaderCount(off, end);
  off = <u32> (hc >>> 32);
  const initialCount = <i32> (hc & 0xffffffff);
  hc = readHeaderCount(off, end);
  off = <u32> (hc >>> 32);
  // operations count consumed but unused (must equal 10,000; parser trusts fixture)

  let childInsertions: u32 = 0;
  let childRemovals: u32 = 0;
  let parentWrites: u32 = 0;
  let nodeCount: u32 = 0;

  // Initial rows: "N\t<id>\t<parentId>\t<position>\t<hexLabel>\n"
  for (let i: i32 = 0; i < initialCount; i++) {
    off += 2; // skip 'N' + '\t'
    let packed = readInt(off, end);
    off = <u32> (packed >>> 32);
    const id = <i32> (packed & 0xffffffff);
    off++; // skip '\t'
    packed = readInt(off, end);
    off = <u32> (packed >>> 32);
    const parentId = <i32> (packed & 0xffffffff);
    off++; // skip '\t'
    packed = readInt(off, end);
    off = <u32> (packed >>> 32);
    const position = <i32> (packed & 0xffffffff);
    off++; // skip '\t'
    const labelOff = off;
    off = readHexSpanEnd(off, end);
    const labelHexLen = <u16> (off - labelOff);
    off = skipLine(off, end);

    setLabelOff(id, labelOff);
    setLabelHexLen(id, labelHexLen);
    if (parentId == -1) {
      setParent(id, -1);
    } else {
      setParent(id, parentId);
      insertAtPosition(parentId, position, id);
      childInsertions++;
      parentWrites++;
    }
    nodeCount++;
  }

  let inserts: u32 = 0;
  let deletes: u32 = 0;
  let reparents: u32 = 0;

  while (off < end) {
    const tag = fixtureAt(off);
    if (tag == 0x0a /* '\n' */) {
      off++;
      continue;
    }
    off++; // skip tag
    off++; // skip '\t'
    if (tag == 0x49 /* 'I' */) {
      let packed = readInt(off, end);
      off = <u32> (packed >>> 32);
      const id = <i32> (packed & 0xffffffff);
      off++;
      packed = readInt(off, end);
      off = <u32> (packed >>> 32);
      const parentId = <i32> (packed & 0xffffffff);
      off++;
      packed = readInt(off, end);
      off = <u32> (packed >>> 32);
      const position = <i32> (packed & 0xffffffff);
      off++;
      const labelOff = off;
      off = readHexSpanEnd(off, end);
      const labelHexLen = <u16> (off - labelOff);
      off = skipLine(off, end);

      setLabelOff(id, labelOff);
      setLabelHexLen(id, labelHexLen);
      setParent(id, parentId);
      insertAtPosition(parentId, position, id);
      inserts++;
      childInsertions++;
      parentWrites++;
      nodeCount++;
    } else if (tag == 0x44 /* 'D' */) {
      const packed = readInt(off, end);
      off = <u32> (packed >>> 32);
      const id = <i32> (packed & 0xffffffff);
      off = skipLine(off, end);
      spliceOut(id);
      setParent(id, -2);
      deletes++;
      childRemovals++;
      parentWrites++;
      nodeCount--;
    } else if (tag == 0x52 /* 'R' */) {
      let packed = readInt(off, end);
      off = <u32> (packed >>> 32);
      const id = <i32> (packed & 0xffffffff);
      off++;
      packed = readInt(off, end);
      off = <u32> (packed >>> 32);
      const parentId = <i32> (packed & 0xffffffff);
      off++;
      packed = readInt(off, end);
      off = <u32> (packed >>> 32);
      const position = <i32> (packed & 0xffffffff);
      off = skipLine(off, end);
      spliceOut(id);
      setParent(id, parentId);
      insertAtPosition(parentId, position, id);
      reparents++;
      childInsertions++;
      childRemovals++;
      parentWrites++;
    } else {
      off = skipLine(off, end);
    }
  }

  fnvReset();
  dfsMix(0);
  const canonicalFnv = fnv;

  store<u32>(RES_OFFSET, inserts);
  store<u32>(RES_OFFSET + 4, deletes);
  store<u32>(RES_OFFSET + 8, reparents);
  store<u32>(RES_OFFSET + 12, nodeCount);
  store<u32>(RES_OFFSET + 16, childInsertions);
  store<u32>(RES_OFFSET + 20, childRemovals);
  store<u32>(RES_OFFSET + 24, parentWrites);
  store<u32>(RES_OFFSET + 28, canonicalFnv);
  return <i32> nodeCount;
}
